File: packages/han/lib/graphql/types/native-task.ts
/**
 * GraphQL NativeTask type
 *
 * Represents a task from Claude Code's built-in TaskCreate/TaskUpdate tools.
 * This is distinct from Han's MCP metrics tasks.
 */

import { nativeTasks } from '../../db/index.ts';
import { builder } from '../builder.ts';

/**
 * NativeTask type from the database
 */
export interface NativeTaskData {
  id: string;
  sessionId: string;
  messageId: string;
  subject: string;
  description: string | null;
  status: string;
  activeForm: string | null;
  owner: string | null;
  blocks: string | null; // JSON array
  blockedBy: string | null; // JSON array
  createdAt: string;
  updatedAt: string;
  completedAt: string | null;
  lineNumber: number;
}

/**
 * NativeTask status enum
 */
export const NativeTaskStatusEnum = builder.enumType('NativeTaskStatus', {
  description: 'Status of a native Claude Code task',
  values: {
    pending: { value: 'pending', description: 'Task not yet started' },
    in_progress: {
      value: 'in_progress',
      description: 'Task currently being worked on',
    },
    completed: { value: 'completed', description: 'Task finished' },
  } as const,
});

/**
 * NativeTask type ref
 */
const NativeTaskRef = builder.objectRef<NativeTaskData>('NativeTask');

/**
 * NativeTask type implementation
 */
export const NativeTaskType = NativeTaskRef.implement({
  description: "A task from Claude Code's built-in task management system",
  fields: (t) => ({
    id: t.exposeID('id', {
      description: 'Task ID (sequential within session)',
    }),
    sessionId: t.exposeString('sessionId', {
      description: 'Session this task belongs to',
    }),
    messageId: t.exposeString('messageId', {
      description: 'Message ID that last updated this task',
    }),
    subject: t.exposeString('subject', {
      description: 'Brief task title',
    }),
    description: t.exposeString('description', {
      description: 'Detailed task description',
      nullable: true,
    }),
    status: t.field({
      type: NativeTaskStatusEnum,
      description: 'Current task status',
      resolve: (task) => task.status as 'pending' | 'in_progress' | 'completed',
    }),
    activeForm: t.exposeString('activeForm', {
      description: 'Present continuous form shown in spinner',
      nullable: true,
    }),
    owner: t.exposeString('owner', {
      description: 'Task owner (agent name)',
      nullable: true,
    }),
    blocks: t.stringList({
      description: 'Task IDs that this task blocks',
      resolve: (task) => {
        if (!task.blocks) return [];
        try {
          return JSON.parse(task.blocks) as string[];
        } catch {
          return [];
        }
      },
    }),
    blockedBy: t.stringList({
      description: 'Task IDs that block this task',
      resolve: (task) => {
        if (!task.blockedBy) return [];
        try {
          return JSON.parse(task.blockedBy) as string[];
        } catch {
          return [];
        }
      },
    }),
    createdAt: t.exposeString('createdAt', {
      description: 'When the task was created',
    }),
    updatedAt: t.exposeString('updatedAt', {
      description: 'When the task was last updated',
    }),
    completedAt: t.exposeString('completedAt', {
      description: 'When the task was completed',
      nullable: true,
    }),
  }),
});

/**
 * Get native tasks for a session
 */
export async function getNativeTasksForSession(
  sessionId: string
): Promise<NativeTaskData[]> {
  const tasks = await nativeTasks.getForSession(sessionId);
  return tasks.map((t) => ({
    id: t.id,
    sessionId: t.sessionId,
    messageId: t.messageId,
    subject: t.subject,
    description: t.description ?? null,
    status: t.status,
    activeForm: t.activeForm ?? null,
    owner: t.owner ?? null,
    blocks: t.blocks ?? null,
    blockedBy: t.blockedBy ?? null,
    createdAt: t.createdAt,
    updatedAt: t.updatedAt,
    completedAt: t.completedAt ?? null,
    lineNumber: t.lineNumber,
  }));
}

File: packages/han-rs/crates/han-api/src/types/native_task.rs
//! Native task (Claude Code TaskCreate/TaskUpdate) GraphQL type.

use async_graphql::*;
use crate::node::encode_global_id;

/// Native task data.
#[derive(Debug, Clone)]
pub struct NativeTask {
    pub raw_id: String,
    pub session_id: String,
    pub message_id: String,
    pub subject: String,
    pub description: Option<String>,
    pub status: String,
    pub active_form: Option<String>,
    pub owner: Option<String>,
    pub blocks: Option<String>,
    pub blocked_by: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub completed_at: Option<String>,
    pub line_number: i32,
}

#[Object]
impl NativeTask {
    async fn id(&self) -> ID { encode_global_id("NativeTask", &self.raw_id) }
    async fn subject(&self) -> &str { &self.subject }
    async fn description(&self) -> Option<&str> { self.description.as_deref() }
    async fn status(&self) -> &str { &self.status }
    async fn active_form(&self) -> Option<&str> { self.active_form.as_deref() }
    async fn owner(&self) -> Option<&str> { self.owner.as_deref() }
    async fn blocks(&self) -> Option<Vec<String>> {
        self.blocks.as_ref().and_then(|b| serde_json::from_str(b).ok())
    }
    async fn blocked_by(&self) -> Option<Vec<String>> {
        self.blocked_by.as_ref().and_then(|b| serde_json::from_str(b).ok())
    }
    async fn created_at(&self) -> &str { &self.created_at }
    async fn updated_at(&self) -> &str { &self.updated_at }
    async fn completed_at(&self) -> Option<&str> { self.completed_at.as_deref() }
}

impl From<han_db::entities::native_tasks::Model> for NativeTask {
    fn from(m: han_db::entities::native_tasks::Model) -> Self {
        Self {
            raw_id: m.id,
            session_id: m.session_id,
            message_id: m.message_id,
            subject: m.subject,
            description: m.description,
            status: m.status,
            active_form: m.active_form,
            owner: m.owner,
            blocks: m.blocks,
            blocked_by: m.blocked_by,
            created_at: m.created_at,
            updated_at: m.updated_at,
            completed_at: m.completed_at,
            line_number: m.line_number,
        }
    }
}
